### PR TITLE
refactor(localization_error_monitor): refactor diag 

### PR DESCRIPTION
**Pull request by YamatoAndo**
## Description

I refactored localization_error_monitor's diagnostics.

- refactor localization_error_monitor's diagnostics code
- add test code for diagnostics code
- add the diagnostics image  to the readme
- rename diagnostic_aggregator's param
 
And I have consolidated what was originally output separately into one.

**before**

Screenshot from 2023-09-12 18-46-47

Screenshot from 2023-09-12 18-46-56

-----

**after**
diagnostics

## Related links

N/A

## Tests performed

- LSim works well.
- check localization_error_monitor's diagnostics by using `ros2 run rqt_runtime_monitor rqt_runtime_monitor`

## Notes for reviewers

N/A

## Interface changes

N/A

## Effects on system behavior

N/A

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: localization/localization_error_monitor/src/diagnostics.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+  for (const auto & stat : stat_array) {
+    if ((stat.level > 0) && (merged_stat.level > 0)) {
+      if (!merged_stat.message.empty()) {
+        merged_stat.message += "; ";
+        merged_stat.message += stat.message;
+      }
+    } else if (stat.level > merged_stat.level) {
+      merged_stat.message = stat.message;
+    }
```
Comment (KYabuuchi) [inline localization/localization_error_monitor/src/diagnostics.cpp:81]: In the original implementation, the condition that goes into line 80 is equivalent to `stat.level>0`, so it can be rewritten as follows.  Also, replacing 0 with Status::OK is easier to understand.
```suggestion
    if ((stat.level > diagnostic_msgs::msg::DiagnosticStatus::OK)) {
      if (!merged_stat.message.empty()) {
        merged_stat.message += "; ";
      }
      merged_stat.message += stat.message;
    }
```

File: localization/localization_error_monitor/src/diagnostics.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  if (ellipse_size >= error_ellipse_size) {
+    stat.level = diagnostic_msgs::msg::DiagnosticStatus::ERROR;
+    stat.message = "ellipse size along lateral direction is over the expected range";
+  }
+
+  return stat;
+}
+
+diagnostic_msgs::msg::DiagnosticStatus mergeDiagnosticStatus(
+  const std::vector<diagnostic_msgs::msg::DiagnosticStatus> & stat_array)
```
Comment (KYabuuchi) [inline localization/localization_error_monitor/src/diagnostics.cpp:69]: Could you add comments to explain how the merging process will work.
For example,  "The highest level within the stat_array will be reflected in the merged_stat. When all stat_array entries are 'OK,' the message of merged_stat will be empty."

**Review by KYabuuchi — COMMENTED** [review]
This refactoring is basically good. But some parts are difficult to read.
Please check my following comments.

File: localization/localization_error_monitor/src/diagnostics.cpp (hunk 1 above)
Comment (YamatoAndo) [reply localization/localization_error_monitor/src/diagnostics.cpp:81]: You're a genius!

File: localization/localization_error_monitor/src/diagnostics.cpp (hunk 2 above)
Comment (YamatoAndo) [reply localization/localization_error_monitor/src/diagnostics.cpp:71]: fixed https://github.com/autowarefoundation/autoware.universe/pull/4964/commits/16e197d5c188f498e4a567e4a70ad610238ac199

**Comment by KYabuuchi** [comment]
We need to modify [`system/system_error_monitor/config/diagnostic_aggregator/localization.param.yaml`](https://github.com/autowarefoundation/autoware.universe/blob/535537db1ab593aaff931c0866f4c92d05e21695/system/system_error_monitor/config/diagnostic_aggregator/localization.param.yaml#L27-L31) too.

The condition needs to be changed from `contains: [": localization_accuracy"]` to `contains: ["localization_accuracy"]` in order to aggregate the concatenated diagnostics.

**Comment by YamatoAndo** [comment]
@ito-san 
I changed the component name from `localization_error_monitor: localization_accuracy` to `localization: localization_error_monitor`, so I updated the diagnostic_aggregator's param. 
Please review this PR.

**Comment by YamatoAndo** [comment]
@KYabuuchi I have made a change so that when the diagnostic status is "OK," the message "OK" will be displayed.

**Review by SakodaShintaro — APPROVED** [review]
Looks Good To Me

**Review by KYabuuchi — APPROVED** [review]
LGTM

**Comment by KYabuuchi** [comment]
RELATED LINK: https://github.com/autowarefoundation/autoware_launch/pull/565 
(It update default parameter of localization_error_monitor)

**Review by ito-san — APPROVED** [review]
LGTM

**Comment by YamatoAndo** [comment]
@Motsu-san Could you please approve this pr?

File: localization/localization_error_monitor/config/localization_error_monitor.param.yaml [hunk 3]
```diff
@@ -4,4 +4,4 @@
     error_ellipse_size: 1.0
     warn_ellipse_size: 0.8
     error_ellipse_size_lateral_direction: 0.3
-    warn_ellipse_size_lateral_direction: 0.2
+    warn_ellipse_size_lateral_direction: 0.25
```
Comment (Motsu-san) [inline localization/localization_error_monitor/config/localization_error_monitor.param.yaml:7]: This seems to be a **fix**. Why was this changed?

File: localization/localization_error_monitor/src/node.cpp [hunk 4]
```diff
@@ -40,7 +41,7 @@ LocalizationErrorMonitor::LocalizationErrorMonitor()
   error_ellipse_size_lateral_direction_ =
     this->declare_parameter("error_ellipse_size_lateral_direction", 0.3);
   warn_ellipse_size_lateral_direction_ =
-    this->declare_parameter("warn_ellipse_size_lateral_direction", 0.2);
+    this->declare_parameter("warn_ellipse_size_lateral_direction", 0.25);
```
Comment (Motsu-san) [inline localization/localization_error_monitor/src/node.cpp:44]: Same as [this](https://github.com/autowarefoundation/autoware.universe/pull/4964/files#diff-9ea4b7dd951014d6499494a7e340fca96ec106bc747f525e655f6acbe94d8954R7).

**Review by Motsu-san — COMMENTED** [review]
I would like to understand why this change is needed and how influences on localization_error_monitor.

**Comment by Motsu-san** [comment]
@YamatoAndo After you resolve my question and conflicts, I will approve.

File: localization/localization_error_monitor/config/localization_error_monitor.param.yaml (hunk 3 above)
Comment (YamatoAndo) [reply localization/localization_error_monitor/config/localization_error_monitor.param.yaml:7]: To be the same value as [this](https://github.com/autowarefoundation/autoware_launch/blob/main/autoware_launch/config/localization/localization_error_monitor.param.yaml#L7).

Comment (Motsu-san) [reply localization/localization_error_monitor/config/localization_error_monitor.param.yaml:7]: Do you mean 0.25 is right? Or is this match with autoware_launch important and can it be changed whether 0.25 or whatever as users like ?

Comment (Motsu-san) [reply localization/localization_error_monitor/config/localization_error_monitor.param.yaml:7]: This is the latter. I got the reply directly.

**Review by Motsu-san — APPROVED** [review]
[LGTMeow](https://lgtmeow.com)
